feature/TECH 507 argo cd support

SamTheisens opened the pull request:
Resolves #181

- [x] Introduce plain k8s manifests as deploy result. These can be deployed with `kubectly apply -f`, or feed into argocd.
- [x] Replace brittle dictionary fiddling
- [x] Move type definitions local to their respective domains
- [x] Clearly describe the 4 different targets of a k8s deploy in documentation 

⚠️ `mpyl_config.yml` will need to have the imagePullSecrets defined like such:
```yaml
    kubernetes:
      imagePullSecrets:
        - name: 'acme-registry'
```


----
### 📕 [TECH-507](https://vandebron.atlassian.net/browse/TECH-507) Integrate MPyL with ArgoCD <img src="https://avatar-management--avatars.us-west-2.prod.public.atl-paas.net/557058:73eb6738-a8dc-4e71-beb2-16761407e54e/44a3caa2-b498-4ee1-927d-bdb0901a683e/24" width="24" height="24" alt="sam@vandebron.nl" /> 
[181](https://github.com/Vandebron/mpyl/issues/181) 

🏗️ Build [17](https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline%20-%20Test/job/PR-185/17/display/redirect) ✅ Successful, started by _Sam Theisens_  
🚀 *[cloudfront-service](https://cloudfront-service-185.test.nl/)*, *job*, *kong-sync*, *[nodeservice](https://nodeservice-185.test.nl/)*, *[sbtservice](https://sbtservice-185.test.nl/)*, *sparkJob*  


[TECH-507]: https://vandebron.atlassian.net/browse/TECH-507?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

koeves commented on src/mpyl/reporting/formatting/markdown.py:
```diff
@@ -81,17 +82,20 @@ def markdown_for_stage(run_result: RunResult, stage: Stage):
+    test_artifacts: dict[str, JunitTestSpec] = _collect_test_artifacts(step_results)
+    test_results: dict[str, list[TestSuite]] = _collect_test_results(test_artifacts)
+
+    listoflists: list[list[TestSuite]] = [value for key, value in test_results.items()]
+    suites: list[TestSuite] = list(itertools.chain.from_iterable(listoflists))
```
I would do this in one line

SamTheisens replied:
Resolved in https://github.com/Vandebron/mpyl/pull/185/commits/383c76e60e9b31de361c83c6a9fa445a39474ab2

koeves commented on src/mpyl/reporting/formatting/markdown.py:
```diff
@@ -120,51 +124,48 @@ def execution_plan_as_markdown(run_result):
 
 
 def to_markdown_test_report(suites: list[TestSuite]):
+    print(suites)
```
Debug print

SamTheisens replied:
Resolved in https://github.com/Vandebron/mpyl/pull/185/commits/383c76e60e9b31de361c83c6a9fa445a39474ab2

Jorg88 commented on README-usage.md:
```diff
@@ -53,7 +54,7 @@ or `lint`.
 [schema](https://vandebron.github.io/mpyl/schema/mpyl_config.schema.yml).  
 Which configuration fields need to be set depends on your usecase. The error messages that you
-encounter while using the cli may guide you through the process.
+may encounter while using the will guide you through the process.
```
```suggestion
may encounter while using the cli will guide you through the process.
```

Jorg88 commented on README.md:
```diff
@@ -36,15 +36,15 @@ MPyL is built with the following principles in mind.
-  outcome. MPyL uses `YAML` strictly for configuration. Where `YAML` needs to produced dynamically, like for CRDs via
-  HELM charts, it is generated by Python code. Te produced `YAML` is validated against the corresponding schemas.
+  outcome. MPyL uses `YAML` strictly for configuration. Where `YAML` needs to be produced dynamically, like for CRD  
+  manifests via HELM charts, it is generated by Python code. Te resulting `YAML` is then validated against the corresponding schemas.
```
```suggestion
  manifests via HELM charts, it is generated by Python code. The resulting `YAML` is then validated against the corresponding schemas.
```

Jorg88 commented on pyproject.toml:
```diff
@@ -0,0 +1,3 @@
+[tool.black]
+line-length = 88
+target-version = ['py39']
```
Actually, should we implement an 'newline at end of the file' linting rule or something? I see quite some of these changes based on the IDE settings people use

SamTheisens replied:
I don't understand how that's possible. We have this rule and it trips the build too.
<img width="408" alt="Screenshot 2023-08-27 at 14 51 36" src="https://github.com/Vandebron/mpyl/assets/1911436/bdc7e5b6-fada-465e-b8c7-a6ab5e5dc4f3">

Jorg88 commented on src/mpyl/reporting/formatting/markdown.py:
```diff
@@ -124,47 +128,43 @@ def to_markdown_test_report(suites: list[TestSuite]):
-def _collect_test_artifacts(step_results: list[StepResult]) -> list[Artifact]:
-    return [
-        res.output.produced_artifact
+def _collect_test_artifacts(step_results: list[StepResult]) -> dict[str, JunitTestSpec]:
```
we should maybe rename the function as well, although technically it's still a test artifact anyway I guess

SamTheisens replied:
https://github.com/Vandebron/mpyl/pull/185/commits/980d7053973d46fa83a93bc5aad4ccdc4f364371

Jorg88 commented on src/mpyl/steps/deploy/k8s/__init__.py:
```diff
@@ -57,7 +105,30 @@ def upsert_namespace(
+    return manifest_file
+
+
+def deploy_helm_chart(  # pylint: disable=too-many-locals
```
This function now does 3 different things, should we maybe split it up?

SamTheisens replied:
I see what you mean. I wouldn't want to increase the burden on the calling side though. It's sort of a strategy pattern without the polymorphic class. We could use that approach, but perhaps a bit over the top.
I've extracted some methods at least

Jorg88 reviewed (commented):
Looks nice, just some minor comments & questions!

I do think it would have been easier for us to review if all the type(casting) changes would have been its own pr. Then it would also be clearer which changes are actually required for argo for us 'noobs' ;-)

SamTheisens commented:
> Looks nice, just some minor comments & questions!
> 
> I do think it would have been easier for us to review if all the type(casting) changes would have been its own pr. Then it would also be clearer which changes are actually required for argo for us 'noobs' ;-)

Yeah, maybe I should have done that. Although it is a relevant refactoring now that we have 4 different deploy artifacts.
More PRs also means a bigger testing burden.
I'm happy to make more PRs, let me know what you prefer.

koeves reviewed (approved):
Looks good! I also would prefer smaller but more PRs, they make for easier and quicker reviews

koeves commented on src/mpyl/schema/project.schema.yml:
```diff
@@ -773,6 +773,15 @@ definitions:
         additionalProperties: false
       portMappings:
         type: object
+      imagePullSecrets:
+        minItems: 1
+        description: 'ImagePullSecrets is a list of references to secrets in the same namespace
+          to use for pulling any images in pods that reference this ServiceAccount. ImagePullSecrets
+          are distinct from Secrets because Secrets can be mounted in the pod, but ImagePullSecrets
+          are only accessed by the kubelet. More info: https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod'
+        items:
+          "$ref": k8s_api_core.schema.yml#/definitions/io.k8s.api.core.v1.LocalObjectReference
+        type: [ array, null ]
```
Shouldn't this be part of the mpyl_config schema?

SamTheisens replied:
You can configure this per project and in `mpyl_config.yml` you can set the defaults. That's why `mpyl_config.yml` embeds parts of the `project.yml` schema.

koeves replied:
Yes but I just tried it and the mpyl config schema validation fails because of the unexpected imagePullSecrets field

SamTheisens replied:
Where did the schema validation fial? The unittests validate the schema here https://github.com/Vandebron/mpyl/blob/2e179fcc4db22dac3b2563a2804bcbfaa68c1a0e/tests/test_resources/mpyl_config.yml#L90-L92

koeves replied:
I added the image pull secrets to the mpyl config, added a dummy change to a project then used this test version and the build failed with 

```
[14:28:50] Schema validation failed Additional properties are not allowed ('imagePullSecrets' was unexpected) at `kubernetes`                                                                           
           Unexpected exception: Additional properties are not allowed ('imagePullSecrets' was unexpected)   
```

it fails at https://github.com/Vandebron/mpyl/blob/b450a7030cb14c4465bae517e034690e628c6584/src/mpyl/build.py#L107-L118 when constructing the `Steps` object (when the mpyl_config.yml validation happens)

koeves replied:
ah maybe I used the wrong place to put the config

koeves replied:
yep wrong kubernetes section :/

koeves commented:
Another weird thing I noticed: I corrected the config so now that part is fine, and in the build plan everything seems fine:
```
+ pipenv run run
🏗️ Building                                                                                                                                                                                              
🏗️ tradingMockServiceApi                                                                                                                                                                                 
📋 tradingMockServiceApi                                                                                                                                                                                
🚀 tradingMockServiceApi  
```

but after the test step it seems like deploy step was skipped completely! I checked in kubernetes and indeed the service was not deployed. In the logs I don't see the usual helm upgrade output either.

I attached the full logs
[Uploading console.txt…]()

SamTheisens commented:
> > Another weird thing I noticed: I corrected the config so now that part is fine, and in the build plan everything seems fine:
> > ```
> > + pipenv run run
> > 🏗️ Building                                                                                                                                                                                              
> > 🏗️ tradingMockServiceApi                                                                                                                                                                                 
> > 📋 tradingMockServiceApi                                                                                                                                                                                
> > 🚀 tradingMockServiceApi  
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but after the test step it seems like deploy step was skipped completely! I checked in kubernetes and indeed the service was not deployed. In the logs I don't see the usual helm upgrade output either.
> > I attached the full logs [console.txt](https://github.com/Vandebron/mpyl/files/12487432/console.txt)
> > this was the run: https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline/job/PR-11303/11/console
> 
> Ah, that doesn't sound good. I'll have a look tomorrow.

It looks like there was/is an issue with the logging. It seems as if log level is set to too high. Only the straight to stdout output is shown.

I ran another build and things seem to work as they should. I hope it's not an intermittent problem. I've added a console print with the log level. So next time this occurs at least we know if there's something wrong with that. 
Could you maybe try another run?

ewaldschrader commented:
Same goes for me concerning the smaller PRs, other than that looks good! :)

koeves commented:
> > > Another weird thing I noticed: I corrected the config so now that part is fine, and in the build plan everything seems fine:
> > > ```
> > > + pipenv run run
> > > 🏗️ Building                                                                                                                                                                                              
> > > 🏗️ tradingMockServiceApi                                                                                                                                                                                 
> > > 📋 tradingMockServiceApi                                                                                                                                                                                
> > > 🚀 tradingMockServiceApi  
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > but after the test step it seems like deploy step was skipped completely! I checked in kubernetes and indeed the service was not deployed. In the logs I don't see the usual helm upgrade output either.
> > > I attached the full logs [console.txt](https://github.com/Vandebron/mpyl/files/12487432/console.txt)
> > > this was the run: https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline/job/PR-11303/11/console
> > 
> > 
> > Ah, that doesn't sound good. I'll have a look tomorrow.
> 
> It looks like there was/is an issue with the logging. It seems as if log level is set to too high. Only the straight to stdout output is shown.
> 
> I ran another build and things seem to work as they should. I hope it's not an intermittent problem. I've added a console print with the log level. So next time this occurs at least we know if there's something wrong with that. Could you maybe try another run?

I checked again with version `185.2368` and same is printed for me (nothing for deployment). Also I checked in kubernetes and the pod of my service was deployed yesterday the last time, so I don't think it's just the logging that's acting up

SamTheisens commented:
> > > > Another weird thing I noticed: I corrected the config so now that part is fine, and in the build plan everything seems fine:
> > > > ```
> > > > + pipenv run run
> > > > 🏗️ Building                                                                                                                                                                                              
> > > > 🏗️ tradingMockServiceApi                                                                                                                                                                                 
> > > > 📋 tradingMockServiceApi                                                                                                                                                                                
> > > > 🚀 tradingMockServiceApi  
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > but after the test step it seems like deploy step was skipped completely! I checked in kubernetes and indeed the service was not deployed. In the logs I don't see the usual helm upgrade output either.
> > > > I attached the full logs [console.txt](https://github.com/Vandebron/mpyl/files/12487432/console.txt)
> > > > this was the run: https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline/job/PR-11303/11/console
> > > 
> > > 
> > > Ah, that doesn't sound good. I'll have a look tomorrow.
> > 
> > 
> > It looks like there was/is an issue with the logging. It seems as if log level is set to too high. Only the straight to stdout output is shown.
> > I ran another build and things seem to work as they should. I hope it's not an intermittent problem. I've added a console print with the log level. So next time this occurs at least we know if there's something wrong with that. Could you maybe try another run?
> 
> I checked again with version `185.2368` and same is printed for me (nothing for deployment). Also I checked in kubernetes and the pod of my service was deployed yesterday the last time, so I don't think it's just the logging that's acting up.
> 
> This is the run: https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline/job/PR-11303/13/console Console: [console.txt](https://github.com/Vandebron/mpyl/files/12497853/console.txt)

There's definitely something wrong with the logging. In the logs you shared none of the mpyl logging is visible. 
It does say:
```
MPyL log level is set to INFO
```
So an incorrect log level is probably not causing it.

I don't have access to Jenkins, so I could only capture logs from the CLI:
[argocd-support-logs.txt](https://github.com/Vandebron/mpyl/files/12503198/argocd-support-logs.txt)

This explains why the services weren't deployed to the cluster:
```
INFO     mpyl.KubernetesDeploy  Service sbtservice reachable at: https://sbtservice-185.test.nl
INFO     mpyl  Execution of Kubernetes Deploy succeeded for 'sbtservice' with outcome 'Wrote kubectl manifest to tests/projects/sbt-service/target/kubernetes' 
```

What is the action configured here https://github.com/Vandebron/mpyl/blob/2e179fcc4db22dac3b2563a2804bcbfaa68c1a0e/mpyl_config.example.yml#L70 in the config repo?

SamTheisens commented:
Reapplied changes in https://github.com/Vandebron/mpyl/pull/185 to merge with main

SamTheisens commented:
Reapplied changes in https://github.com/Vandebron/mpyl/pull/185 to merge with main
